### PR TITLE
142.Linked List Cycle II

### PR DESCRIPTION
**Pull request by tk-hirom**
## 問題

https://leetcode.com/problems/linked-list-cycle-ii/description/

## 問題文

Given the head of a linked list, return the node where the cycle begins. If there is no cycle, return null.

There is a cycle in a linked list if there is some node in the list that can be reached again by continuously following the next pointer. Internally, pos is used to denote the index of the node that tail's next pointer is connected to (0-indexed). It is -1 if there is no cycle. Note that pos is not passed as a parameter.

Do not modify the linked list.

 

Example 1:

Input: head = [3,2,0,-4], pos = 1
Output: tail connects to node index 1
Explanation: There is a cycle in the linked list, where tail connects to the second node.
Example 2:

Input: head = [1,2], pos = 0
Output: tail connects to node index 0
Explanation: There is a cycle in the linked list, where tail connects to the first node.
Example 3:

Input: head = [1], pos = -1
Output: no cycle
Explanation: There is no cycle in the linked list.
 

Constraints:

The number of the nodes in the list is in the range [0, 104].
-105 <= Node.val <= 105
pos is -1 or a valid index in the linked-list.

### CODE REVIEW

File: 142. Linked List Cycle II/answer.md [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+fastは2倍の速度で動いていたから、2(T+C)と表せる。またfastはslowと出会うまでサイクルの中をk周したとすると、kL+T+Cと表せる
+
+すると、kl = T+Cと言える。というところまでは理解できたが、それが何を指しているのかがわからない。
+
+衝突地点からslow, fastに来た道を１ステップずつ戻ってもらう
+
+同じ速度で戻るからサイクルのスタート地点までは同伴している。（Cの距離分一緒に戻った）
+
+その後Tの距離分slowが戻る。するとfastは衝突地点に戻っている。なぜならT+C=kLで亀の移動分動かすと、サイクルを何周かしたのと同じになるからちょうどスタートポジションである衝突地点に戻る
```
Comment (oda) [inline 142. Linked List Cycle II/answer.md:89]: 私の説明。
https://discord.com/channels/1084280443945353267/1246383603122966570/1252209488815984710

Comment (tk-hirom) [reply 142. Linked List Cycle II/answer.md:89]: 説明リンク追加ありがとうございます！mdにも書いておきます！

File: 142. Linked List Cycle II/answer.md [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+#         self.val = x
+#         self.next = None
+
+class Solution:
+    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
+        collision = self.findCollision(head)
+        if not collision:
+            return None
+
+        fromCollision = collision
```
Comment (TORUS0818) [inline 142. Linked List Cycle II/answer.md:104]: PEP8貼っておきます。
https://peps.python.org/pep-0008/#function-and-variable-names

File: 142. Linked List Cycle II/answer.md [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+            fromCollision = fromCollision.next
+        return fromHead
+        
+
+    def findCollision(self, head) -> Optional[ListNode]:
+        fast = slow = head
+        while fast and fast.next:
+            fast = fast.next.next
+            slow = slow.next
+            if(fast == slow):
```
Comment (TORUS0818) [inline 142. Linked List Cycle II/answer.md:118]:
```python
if fast == slow:
```
で良いですね。

Comment (tk-hirom) [reply 142. Linked List Cycle II/answer.md:118]: 他の言語の癖でつい書いちゃいました💦
気をつけます

**Pull request merged**